backend/de.metas.contracts/src/main/java/de/metas/contracts/modular/impl/SalesOrderLineModularContractHandler.java

```diff
@@ -0,0 +1,267 @@
+
+		return modularContractTypeId.map(contractTypeId -> LogEntryCreateRequest.builder()
+				.referencedRecord(TableRecordReference.of(I_C_OrderLine.Table_Name, orderLine.getC_OrderLine_ID()))
+				.contractId(modularContractId)
+				.collectionPointBPartnerId(bPartnerId)
+				.producerBPartnerId(bPartnerId)
+				.invoicingBPartnerId(bPartnerId)
+				.warehouseId(WarehouseId.ofRepoId(order.getM_Warehouse_ID()))
+				.productId(productId)
+				.documentType(LogEntryDocumentType.SALES_ORDER)
+				.soTrx(SOTrx.PURCHASE)
+				.processed(false)
+				.quantity(quantity)
+				.transactionDate(LocalDateAndOrgId.ofTimestamp(order.getDateOrdered(),
+															   OrgId.ofRepoId(orderLine.getAD_Org_ID()),
+															   orgDAO::getTimeZone))
+				.year(modularContractSettings.getYearAndCalendarId().yearId())
+				.description(description)
+				.modularContractTypeId(contractTypeId)
+				.build());
+	}
+
+	@Override
+	public @NonNull Optional<LogEntryReverseRequest> createLogEntryReverseRequest(final @NonNull I_C_OrderLine orderLine, final @NonNull FlatrateTermId flatrateTermId)
+	{
+		final I_C_UOM uomId = uomDAO.getById(UomId.ofRepoId(orderLine.getC_UOM_ID()));
+		final Quantity quantity = Quantity.of(orderLine.getQtyEntered(), uomId);
+
+		final String description = msgBL.getMsg(MSG_ON_REVERSE_DESCRIPTION, ImmutableList.of(String.valueOf(orderLine.getM_Product_ID()), quantity.toString()));
+
+		return Optional.of(LogEntryReverseRequest.builder()
+								   .referencedModel(TableRecordReference.of(I_C_OrderLine.Table_Name, orderLine.getC_OrderLine_ID()))
+								   .flatrateTermId(flatrateTermId)
+								   .description(description)
+								   .build());
+	}
+
+	@Override
+	public @NonNull Stream<FlatrateTermId> streamContractIds(@NonNull final I_C_OrderLine orderLine)
+	{
+		final I_C_Order order = orderBL.getById(OrderId.ofRepoId(orderLine.getC_Order_ID()));
+		final BPartnerLocationAndCaptureId orderBillPartnerId = orderBL.getBillToLocationId(order);
+		final YearId harvestingYearId = YearId.ofRepoIdOrNull(order.getHarvesting_Year_ID());
+
+		Check.assume(harvestingYearId != null, "Harvesting year ID should not be null at this stage!");
+
+		final ModularFlatrateTermRequest request = ModularFlatrateTermRequest.builder()
+				.bPartnerId(orderBillPartnerId.getBpartnerId())
+				.productId(ProductId.ofRepoId(orderLine.getM_Product_ID()))
+				.yearId(harvestingYearId)
+				.soTrx(SOTrx.PURCHASE)
+				.build();
+
+		return streamModularContracts(request)
+				.map(I_C_Flatrate_Term::getC_Flatrate_Term_ID)
+				.map(FlatrateTermId::ofRepoId);
+	}
+
+	@Override
+	public void validateDocAction(final @NonNull I_C_OrderLine model, final ModularContractService.@NonNull ModelAction action)
+	{
+		switch (action)
+		{
+			case COMPLETED, VOIDED, REVERSED, REACTIVATED -> {}
+			case DELETED -> validateCanBeDeleted(model);
```

Comment:
IMO DELETE is not a docAction.
I would move all the validation for the deleted line in a separate model interceptor, with 	``@ModelChange(timings = ModelValidator.TYPE_BEFORE_DELETE)``